File: demo/web/scripts/prepare_demo.js
const fs = require("fs");
const path = require("path");
const testData = require("../../../resources/.test/test_data.json");

availableLanguages = testData["tests"]["within_context"].map(
  (x) => x["language"]
);

const language = process.argv.slice(2)[0];
if (language === "==") {
  console.error(`Choose the language you would like to run the demo in with "yarn start [language]". 
        Available languages are ${availableLanguages.join(", ")}`);
  process.exit(1);
}

if (!availableLanguages.includes(language)) {
  console.error(`'${language}' is not an available demo language. 
        Available languages are ${availableLanguages.join(", ")}`);
  process.exit(1);
}

const context = testData["tests"]["within_context"].find(
  (x) => x["language"] === language
)["context_name"];
const contextFileName = `${context}_wasm.rhn`;

const suffix = language === "en" ? "" : `_${language}`;
const rootDir = path.join(__dirname, "..", "..", "..");

const contextDir = path.join(rootDir, "resources", `contexts${suffix}`, "wasm");

let outputDirectory = path.join(__dirname, "..", "contexts");
if (fs.existsSync(outputDirectory)) {
  fs.readdirSync(outputDirectory).forEach((k) => {
    fs.unlinkSync(path.join(outputDirectory, k));
  });
} else {
  fs.mkdirSync(outputDirectory, { recursive: true });
}

try {
  fs.copyFileSync(
    path.join(contextDir, contextFileName),
    path.join(outputDirectory, contextFileName)
  );
} catch (error) {
  console.error(error);
  process.exit(1);
}

fs.writeFileSync(
  path.join(outputDirectory, "rhinoContext.js"),
  `const rhinoContext = {
  publicPath: "contexts/${contextFileName}",
  forceWrite: true,
};

(function () {
  if (typeof module !== "undefined" && typeof module.exports !== "undefined")
    module.exports = rhinoContext;
})();`
);

const modelDir = path.join(rootDir, "lib", "common");

outputDirectory = path.join(__dirname, "..", "models");
if (fs.existsSync(outputDirectory)) {
  fs.readdirSync(outputDirectory).forEach((k) => {
    fs.unlinkSync(path.join(outputDirectory, k));
  });
} else {
  fs.mkdirSync(outputDirectory, { recursive: true });
}

const modelName = `rhino_params${suffix}.pv`;
fs.copyFileSync(
  path.join(modelDir, modelName),
  path.join(outputDirectory, modelName)
);

fs.writeFileSync(
  path.join(outputDirectory, "rhinoModel.js"),
  `const rhinoModel = {
  publicPath: "models/${modelName}",
  forceWrite: true,
};

(function () {
  if (typeof module !== "undefined" && typeof module.exports !== "undefined")
    module.exports = rhinoModel;
})();`
);

File: demo/flutter/scripts/prepare_demo.dart
import "package:path/path.dart";

import "dart:convert";
import "dart:io";

final String resourcePath =
    join(dirname(Platform.script.path), "..", "..", "..", "resources");
final String libPath =
    join(dirname(Platform.script.path), "..", "..", "..", "lib");
final String testDataPath = join(resourcePath, ".test", "test_data.json");

final String assetsPath = join(dirname(Platform.script.path), "..", "assets");
final String contextsPath = join(assetsPath, "contexts");
final String modelsPath = join(assetsPath, "models");

Future<Map> readJsonFile(String filePath) async {
  var input = await File(filePath).readAsString();
  var map = jsonDecode(input);
  return map;
}

void main(List<String> arguments) async {
  var testData = await readJsonFile(testDataPath);
  List<String> availableLanguages = List<String>.from(
      testData["tests"]["within_context"].map((x) => x["language"]).toList());

  if (arguments.isEmpty) {
    print(
        "Choose the language you would like to run the demo in with 'dart scripts/prepare_demo.dart [language]'.\n"
        "Available languages are ${availableLanguages.join(", ")}.");
    exit(1);
  }

  String language = arguments[0];
  String suffix = (language == "en") ? "" : "_$language";
  if (!availableLanguages.contains(language)) {
    print("'$language' is not an available demo language.\n"
        "Available languages are ${availableLanguages.join(", ")}.");
    exit(1);
  }

  String contextName = testData["tests"]["within_context"]
      .firstWhere((x) => x["language"] == language)["context_name"];

  var androidContextsDirSrc =
      Directory(join(resourcePath, "contexts$suffix", "android"));
  var iOSContextsDirSrc =
      Directory(join(resourcePath, "contexts$suffix", "ios"));

  var androidContextsDirDst = Directory(join(contextsPath, 'android'));
  if (androidContextsDirDst.existsSync()) {
    androidContextsDirDst.deleteSync(recursive: true);
  }
  androidContextsDirDst.createSync(recursive: true);

  var iOSContextsDirDst = Directory(join(contextsPath, 'ios'));
  if (iOSContextsDirDst.existsSync()) {
    iOSContextsDirDst.deleteSync(recursive: true);
  }
  iOSContextsDirDst.createSync(recursive: true);

  var modelDir = Directory(modelsPath);
  if (modelDir.existsSync()) {
    modelDir.deleteSync(recursive: true);
  }
  modelDir.createSync(recursive: true);

  var params = <String, String>{};
  params["language"] = language;
  params["context"] = contextName;

  File androidContextSrc =
      File(join(androidContextsDirSrc.path, "${contextName}_android.rhn"));
  androidContextSrc.copySync(
      join(androidContextsDirDst.path, basename(androidContextSrc.path)));

  File iOSContextSrc =
      File(join(iOSContextsDirSrc.path, "${contextName}_ios.rhn"));
  iOSContextSrc
      .copySync(join(iOSContextsDirDst.path, basename(iOSContextSrc.path)));

  if (language != "en") {
    File model = File(join(libPath, "common", "rhino_params$suffix.pv"));
    model.copySync(join(modelDir.path, basename(model.path)));
  }

  var encoded = json.encode(params);
  File f = File(join(assetsPath, "params.json"));
  f.writeAsStringSync(encoded);

  print("Demo is ready to run!");
}
